src/lib.rs

```diff
@@ -53,21 +54,73 @@ impl Command {
         }
     }
 
-    /// Like [`std::process::Command::arg`][process::Command#method.arg], add an argument to the
-    /// command, but only if it will fit.
-    pub fn try_arg<S: AsRef<OsStr>>(&mut self, arg: S) -> bool {
-        if arg.as_ref().len() as i64 > platform::max_single_argument_length() {
-            return false;
+    /// Return the appropriate error for a too-big command line.
+    fn e2big() -> io::Error {
+        // io::ErrorKind::ArgumentListTooLong is unstable, so get it manually
+        #[cfg(unix)]
+        return io::Error::from_raw_os_error(libc::E2BIG);
+        #[cfg(not(unix))]
+        return io::ErrorKind::Other.into();
+    }
+
+    /// Get the size of some arguments, if they'll fit.
+    fn check_size<I, S>(&self, args: I) -> io::Result<i64>
+    where
+        I: IntoIterator<Item = S>,
+        S: AsRef<OsStr>,
+    {
+        let mut size = 0;
+        for arg in args {
+            let arg = arg.as_ref();
+            if arg.len() as i64 > platform::max_single_argument_length() {
+                return Err(Self::e2big());
+            }
+            size += platform::arg_size(arg);
         }
 
-        let arg_size = platform::arg_size(&arg);
-        if arg_size > self.remaining_argument_length {
-            false
-        } else {
-            self.remaining_argument_length -= arg_size;
-            self.inner.arg(arg);
-            true
+        if size > self.remaining_argument_length {
+            return Err(Self::e2big());
         }
+
+        Ok(size)
+    }
+
+    /// Check if an additional argument would fit in the command.
+    pub fn arg_would_fit<S: AsRef<OsStr>>(&self, arg: S) -> bool {
+        self.args_would_fit(&[arg])
+    }
+
+    /// Check if multiple additional arguments would all fit in the command.
+    pub fn args_would_fit<I, S>(&self, args: I) -> bool
+    where
+        I: IntoIterator<Item = S>,
+        S: AsRef<OsStr>,
+    {
+        self.check_size(args).is_ok()
+    }
+
+    /// Like [`std::process::Command::arg`][process::Command#method.arg], add an argument to the
+    /// command, but only if it will fit.
+    pub fn try_arg<S: AsRef<OsStr>>(&mut self, arg: S) -> io::Result<&mut Self> {
+        let arg = arg.as_ref();
+        let size = self.check_size(&[arg])?;
+        self.inner.arg(arg);
+        self.remaining_argument_length -= size;
+        Ok(self)
+    }
+
+    /// Like [`std::process::Command::arg`][process::Command#method.args], add multiple arguments to
+    /// the command, but only if they will all fit.
+    pub fn try_args<I, S>(&mut self, args: I) -> io::Result<&mut Self>
+    where
+        I: IntoIterator<Item = S>,
+        S: AsRef<OsStr>,
+    {
+        let args = Vec::from_iter(args);
+        let size = self.check_size(&args)?;
+        self.inner.args(&args);
```

Comment:
Minor point: can we avoid the creation of the `Vec` here? Maybe by putting additional constraints on `I` or `S`. Or by not using `IntoIterator` but `Iterator`? Or do we rather want to keep consistency with the `std::process::Command::args`method here?

Reply:
I thought `for<'a> &'a I: IntoIterator<Item = S>` would work but it doesn't.  `IntoIterator<Item = S> + Copy` might work.

Reply:
Yeah this works:

```diff
diff --git a/src/lib.rs b/src/lib.rs
index 5b868e7..7219d03 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -20,7 +20,6 @@
 
 use std::ffi::OsStr;
 use std::io;
-use std::iter::FromIterator;
 use std::path::Path;
 use std::process::{self, Child, ExitStatus, Output, Stdio};
 
@@ -102,23 +101,18 @@ impl Command {
     /// Like [`std::process::Command::arg`][process::Command#method.arg], add an argument to the
     /// command, but only if it will fit.
     pub fn try_arg<S: AsRef<OsStr>>(&mut self, arg: S) -> io::Result<&mut Self> {
-        let arg = arg.as_ref();
-        let size = self.check_size(&[arg])?;
-        self.inner.arg(arg);
-        self.remaining_argument_length -= size;
-        Ok(self)
+        self.try_args(&[arg])
     }
 
     /// Like [`std::process::Command::arg`][process::Command#method.args], add multiple arguments to
     /// the command, but only if they will all fit.
     pub fn try_args<I, S>(&mut self, args: I) -> io::Result<&mut Self>
     where
-        I: IntoIterator<Item = S>,
+        I: IntoIterator<Item = S> + Copy,
         S: AsRef<OsStr>,
     {
-        let args = Vec::from_iter(args);
-        let size = self.check_size(&args)?;
-        self.inner.args(&args);
+        let size = self.check_size(args)?;
+        self.inner.args(args);
         self.remaining_argument_length -= size;
         Ok(self)
     }
```

If you like that I'll go ahead with it, I agree the temporary `Vec` is not ideal.

Reply:
I'd prefer that, yes. Unless this puts strange restrictions on how this function can be used. I guess it would still be possible to call it with `.try_args(&["foo", "bar"])`.